File: TNET_UI/Sources/CandidateSet.cpp

#include "CandidateSet.h"

int CandidateSet::getCount()
{
	return _transactions.size();
}

std::hash_map<Hash, TransactionContent>  CandidateSet::getTransactions()
{
	return _transactions;
}

CandidateSet::CandidateSet()
{
	//_transactions;
}

void CandidateSet::AddTransaction(TransactionContent transaction)
{
	_transactions[transaction.GetHash()]= transaction;	
}

void CandidateSet::GenerateTransactions(vector<AccountInfo> accounts)
{
	_transactions.clear();

	int totalAccounts = (int)accounts.size();
	
	std::default_random_engine generator;

	generator.seed(rand());

	std::uniform_int_distribution<int> distribution(1, totalAccounts / 5);
	std::uniform_int_distribution<int> distribution_total(0, totalAccounts-1);

	double lower_bound = 0;
	double upper_bound = 1;
	std::uniform_real_distribution<double> unif(lower_bound, upper_bound);

	std::default_random_engine re;

	re.seed(rand());

	for (int i = 0; i < (int)accounts.size();i++)
	{
		AccountInfo account = accounts[i];

		int nodes = distribution(generator);

		long amountToSpend = (long)(account.Money * 1.5F);

		long spendingAmt = (amountToSpend / (long)nodes);

		vector<TransactionSink> tsks;

		for (int dest = 0; dest < nodes; dest++)
		{
			int _id = distribution_total(generator);
			AccountInfo sinkAccount = accounts[_id];

			if (sinkAccount.AccountID == account.AccountID) continue;
			double d = unif(re);
			long amountToSpendPerSink = (long)(spendingAmt * d);

			TransactionSink tsk = TransactionSink(sinkAccount.AccountID, amountToSpendPerSink);
			tsks.push_back(tsk);
		}

		TransactionContent tco =  TransactionContent(account.AccountID, 0, tsks, Hash() );
		Hash h = tco.GetHash();
		_transactions[h] = tco;
	}

}

File: tnetd/Consensus/CandidateSet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TNetD.Transactions;

namespace TNetD.Consensus
{

    class CandidateSet
    {
        List<TransactionContent> _transactions;

        public TransactionContent this[int index]
        {
            get
            {
                return _transactions[index];
            }
        }

        public int Count
        {
            get { return _transactions.Count; }
        }

        public List<TransactionContent> Transactions
        {
            get { return _transactions; }
        }

        public CandidateSet()
        {
            _transactions = new List<TransactionContent>();
        }

        public void AddTransaction(TransactionContent transaction)
        {
            _transactions.Add(transaction);
        }

        ///// <summary>
        ///// *******  TEST FUNCTION ONLY  ******
        ///// </summary>
        ///// <param name="accounts"></param>
        //public void GenerateTransactions(List<AccountInfo> accounts)
        //{
        //    _transactions.Clear();

        //    int totalAccounts = accounts.Count;

        //    Random rnd = new Random();

        //    foreach (AccountInfo account in accounts)
        //    {
        //        int nodes = rnd.Next(1, totalAccounts / 5);

        //        long amountToSpend = (long)(account.Money * 1.5F);

        //        long spendingAmt = (amountToSpend / (long)nodes);

        //        List<TransactionSink> tsks = new List<TransactionSink>();

        //        for (int dest = 0; dest < nodes; dest++)
        //        {
        //            int _id = rnd.Next(0, totalAccounts);
        //            AccountInfo sinkAccount = accounts[_id];

        //            if (sinkAccount.AccountID == account.AccountID) continue;

        //            long amountToSpendPerSink = (long)(spendingAmt * rnd.NextDouble()); //(long)rnd.Next(Math.Max((int)spendingAmt/2, 1), Math.Max((int)spendingAmt, 1));

        //            TransactionSink tsk = new TransactionSink(sinkAccount.AccountID, amountToSpendPerSink);
        //            tsks.Add(tsk);
        //        }

        //        TransactionContent tco = new TransactionContent(account.AccountID, 0, tsks.ToArray(), new Hash[0]);
        //        _transactions.Add(tco);
        //    }

        //}

    }

}
